ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs

```diff
@@ -228,6 +234,32 @@ runNodeNetwork registry testBtime numCoreNodes nodeJoinPlan nodeTopology
           simChaChaT varDRG id $ testGenTxs numCoreNodes cfg ledger
         void $ addTxs mempool txs
 
+    ebbProducer :: HasCallStack
+                => StrictTVar m SlotNo
+                -> NodeConfig (BlockProtocol blk)
+                -> ChainDB.ChainDB m blk
+                -> EpochInfo m
+                -> m ()
+    ebbProducer v cfg chainDB epochInfo = go 0
+      where
+        go :: EpochNo -> m ()
+        go !epoch = do
+          -- The first slot in @epoch@
+          ebbSlotNo <- epochInfoFirst epochInfo epoch
+          atomically $ writeTVar v ebbSlotNo
+
+          void $ blockUntilSlot btime ebbSlotNo
+
+          (prevSlot, ebbBlockNo, prevHash) <- ChainDB.getTipHeader chainDB <&> \case
+            Nothing  -> (Origin            , 0,           GenesisHash)
+            Just hdr -> (At (blockSlot hdr), blockNo hdr, BlockHash (headerHash hdr))
+          when (prevSlot < At ebbSlotNo) $ case mbForgeEBB of
+            Nothing       -> pure ()
+            Just forgeEBB -> do
+              let ebb = forgeEBB cfg ebbSlotNo ebbBlockNo prevHash
+              ChainDB.addBlock chainDB ebb
```

Comment:
I just realized that this `addBlock` achieves nothing for epochs after epoch 0.

https://github.com/input-output-hk/ouroboros-network/blob/602856e08e0ebe9d63751185d39580d3b22989c1/ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/ChainSel.hs#L377-L379

So:

  * The ChainDB will only select an EBB if that EBB already has a successor (with an exception for EBB0 since it's predecessor is `Origin`).

But that can never happen since

  * The node only forges blocks onto the current tip of the ChainDB.

(We have neither the chicken nor the egg!)

To test with EBBs in later epochs, we'll need to relax/override one of those bullet points.

Reply:
I may have found another option. See the third commit discussed at https://github.com/input-output-hk/ouroboros-network/pull/1353#issuecomment-566223460